inspirehep/modules/records/api.py

```diff
@@ -90,6 +91,50 @@ def _create_bucket(self, location=None, storage_class=None):
         )
         return bucket
 
+    def attach_file(self, file_name, file):
+        """Attach the file ``file`` to the record's files properties,
+        updating the record's metadata..
+        The file is attached only if there is a key in ``documents`` or
+        ``figures`` which key match ``file_name``.
+        Then. the name of the file is changed adding the record's
+        ``control_number``as prefix. This change is reflected in
+        the record's metadata too.
+        Args:
+            file_name(string): the name of the file to attach. It must be a
+                key of ``documents`` or ``figures``.
+            file(file): the file to attach to the file.
+        Return:
+            (string): the new file's key.
+        Raise:
+            ValueError in case ``file_name`` is not in record's documents and
+                figures.
+        Example:
+            >>with open(my_file) as stream:
+                record.attach_file('NewFile', stream)
+        """
+
+        control_number = self.get('control_number')
+        normalized_file_name = u'{}_{}'.format(control_number, file_name)
+
+        for attachment in itertools.chain(
```

Comment:
this is now quadratic in the number of files + figures. The number of figures can be quite large, so maybe you should have an `attach_files` function that takes no arguments and does this for all files?

Reply:
this depends on how many times you call the `attach_file`. We want to handle in a single point the attachment of a file because we have the control_number logic and we use it in different places. So, what's the problem with this one, sorry I didn't understand it.

Reply:
How large is quite large?

Reply:
(in this case, it iterates the files first, and the images after, so the only resource that is being "wasted" is a bit of memory to hold the list of figures while waiting to iterate over them)

Reply:
hundreds

Reply:
@michamos https://github.com/ammirate/inspire-next/commit/26f9cfa67bdb8656a976efd8af5c66fe73125d50#diff-1bdc7b327ecbacc5d173e9ff21842a8dR91 here we iterate through different dataset and we are trying to be consistent with this too. For this PR is fine to include a resolver but for merger PR we cannot I think.